File: 16/16.js
const fs = require('fs');
const input = fs.readFileSync('16/16.txt', 'UTF-8');

const p1 = (input) => {
    let lines = input.split('\n');

    let threeCount = 0;

    while(true) {
        let before = lines.shift();

        if(!/^Before/.test(before)) {
            break;
        }

        let argString = before.split('[')[1];
        argString = argString.substring(0, argString.length - 1);
        let beforeArgs = argString.split(', ').map(x => +x);

        let opString = lines.shift();
        let opArgs = opString.split(' ').map(x => +x);

        let after = lines.shift();
        argString = after.split('[')[1];
        argString = argString.substring(0, argString.length - 1);
        let afterArgs = argString.split(', ').map(x => +x);
        lines.shift();

        let matchCount = 0;

        let aReg = beforeArgs[opArgs[1]];
        let bReg = beforeArgs[opArgs[2]];
        let aVal = opArgs[1];
        let bVal = opArgs[2];
        let cVal = opArgs[3];

        //addr
        let candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg + bReg;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //addi
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg + bVal;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //mulr
        candidates= beforeArgs.slice(0);
        candidates[cVal] = aReg * bReg;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //muli
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg * bVal;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //banr
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg & bReg;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //bani
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg & bVal;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //borr
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg | bReg;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //bori
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg | bVal;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //setr
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //seti
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aVal;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //gtir
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aVal > bReg ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //gtri
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg > bVal ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //gtrr
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg > bReg ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //eqir
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aVal === bReg ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //eqri
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg === bVal ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        //eqrr
        candidates = beforeArgs.slice(0);
        candidates[cVal] = aReg === bReg ? 1 : 0;
        if(candidates.join(',') === afterArgs.join(',')) {
            matchCount++;
        }

        if(matchCount >= 3) {
            threeCount++;
        }
    }

    return threeCount;
};

const p2 = (input) => {
  const ops = {
    addr: (reg, a, b) => reg[a] + reg[b],
    addi: (reg, a, b) => reg[a] + b,
    mulr: (reg, a, b) => reg[a] * reg[b],
    muli: (reg, a, b) => reg[a] * b,
    banr: (reg, a, b) => reg[a] & reg[b],
    bani: (reg, a, b) => reg[a] & b,
    borr: (reg, a, b) => reg[a] | reg[b],
    bori: (reg, a, b) => reg[a] | b,
    setr: (reg, a, b) => reg[a],
    seti: (reg, a, b) => a,
    gtir: (reg, a, b) => a > reg[b] ? 1 : 0,
    gtri: (reg, a, b) => reg[a] > b ? 1 : 0,
    gtrr: (reg, a, b) => reg[a] >  reg[b] ? 1 : 0,
    eqir: (reg, a, b) => a === reg[b] ? 1 : 0,
    eqri: (reg, a ,b) => reg[a] === b ? 1 : 0,
    eqrr: (reg, a, b) => reg[a] === reg[b] ? 1 : 0
  };

  let lines = input.split('\n');

  let opCandidates = {};

  for(let opCode in ops) {
      opCandidates[opCode] = [];
  }

  while(true) {
      let before = lines.shift();

      if(!/^Before/.test(before)) {
          break;
      }

      let argString = before.split('[')[1];
      argString = argString.substring(0, argString.length - 1);
      let beforeArgs = argString.split(', ').map(x => +x);

      let opString = lines.shift();
      let opArgs = opString.split(' ').map(x => +x);

      let after = lines.shift();
      argString = after.split('[')[1];
      argString = argString.substring(0, argString.length - 1);
      let afterArgs = argString.split(', ').map(x => +x);
      lines.shift();

      for(let opCode in ops) {
          let candidates = beforeArgs.slice(0);
          candidates[opArgs[3]] = ops[opCode](candidates, opArgs[1], opArgs[2]);

          if(candidates.join(',') === afterArgs.join(',')) {
              if(opCandidates[opCode].indexOf(opArgs[0]) === -1) {
                  opCandidates[opCode].push(opArgs[0]);
              }
          }
      }
  }

  let foundInstructions = 0;

  let opCodes = {};

  while(foundInstructions < 16) {
      for(let opCode in opCandidates) {
          if(opCandidates[opCode].length === 1) {
              let foundNum = opCandidates[opCode][0];
              foundInstructions++;
              opCodes[foundNum] = opCode;

              for(let opCode2 in opCandidates) {
                  let index = opCandidates[opCode2].indexOf(foundNum);
                  if(index > -1) {
                      opCandidates[opCode2].splice(index, 1);
                  }
              }
          }
      }

      for(let i = 0; i < 16; i++) {
          let foundCount = 0;
          let foundCode = '';

          for(let opCode in opCandidates) {
              if(opCandidates[opCode].indexOf(i) >= 0) {
                  foundCount++;
                  foundCode = opCode;
              }
          }

          if(foundCount === 1) {
              foundInstructions++;
              opCodes[i] = foundCode;

              for(let opCode in opCandidates) {
                  let index = opCandidates[opCode].indexOf(i);
                  if(index >= 0) {
                      opCandidates[opCode].splice(index, 1);
                  }
              }
              opCandidates[foundCode] = [];
          }
      }
  }

  lines.shift();
  lines.shift();

  let registers = [0, 0, 0, 0];

  while(lines.length) {
      let line = lines.shift();
      let args = line.split(' ').map(x => +x);

      let opCode = opCodes[args[0]];
      registers[args[3]] = ops[opCode](registers, args[1], args[2]);
  }

  return registers[0];
};

console.log('Part one:', p1(input));
console.log('Part two:', p2(input));

File: 16/16.py
import fileinput
import re

# the instructions: r is a set of registers
def addr(r, a, b, c): r[c] = r[a] + r[b]
def addi(r, a, b, c): r[c] = r[a] + b
def mulr(r, a, b, c): r[c] = r[a] * r[b]
def muli(r, a, b, c): r[c] = r[a] * b
def banr(r, a, b, c): r[c] = r[a] & r[b]
def bani(r, a, b, c): r[c] = r[a] & b
def borr(r, a, b, c): r[c] = r[a] | r[b]
def bori(r, a, b, c): r[c] = r[a] | b
def setr(r, a, b, c): r[c] = r[a]
def seti(r, a, b, c): r[c] = a
def gtir(r, a, b, c): r[c] = int(a > r[b])
def gtri(r, a, b, c): r[c] = int(r[a] > b)
def gtrr(r, a, b, c): r[c] = int(r[a] > r[b])
def eqir(r, a, b, c): r[c] = int(a == r[b])
def eqri(r, a, b, c): r[c] = int(r[a] == b)
def eqrr(r, a, b, c): r[c] = int(r[a] == r[b])

functions = [
    addr, addi, mulr, muli, banr, bani, borr, bori,
    setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr,
]

# parse pulls out `opcode, A, B, C`, all integers
def parse(line):
    return list(map(int, re.findall(r'\d+', line)))

# behaves_like indicates how many functions match the observed behavior
# in the before and after registers when running the provided instruction
def behaves_like(instruction, before, after):
    count = 0
    for f in functions:
        r = list(before)
        f(r, *instruction[1:])
        if r == after:
            count += 1
    return count

# remove_candidates discards functions that do not match the observed behavior
def remove_candidates(instruction, before, after, candidates):
    for f in functions:
        r = list(before)
        f(r, *instruction[1:])
        if r != after:
            candidates[instruction[0]].discard(f)

lines = list(fileinput.input())

# part 1
count = 0
for line in lines:
    if 'Before' in line:
        before = parse(line)
    elif 'After' in line:
        after = parse(line)
        # count how many examples behave like 3 or more functions
        if behaves_like(instruction, before, after) >= 3:
            count += 1
    else:
        instruction = parse(line)
print(count)

# part 2A - determine which opcodes map to which functions
opcodes = {}
known = set()
# loop until we've identified all opcodes
while len(known) < len(functions):
    # remaining candidates are the set of all functions
    # minus the set of known functions
    candidates = {}
    for i in range(len(functions)):
        candidates[i] = set(functions) - set(known)
    for line in lines:
        if 'Before' in line:
            before = parse(line)
        elif 'After' in line:
            after = parse(line)
            remove_candidates(instruction, before, after, candidates)
        else:
            instruction = parse(line)
    # look for opcodes that only have one candidate function
    # add them to the `known` set
    for i in range(len(functions)):
        if len(candidates[i]) == 1:
            f = candidates[i].pop()
            opcodes[i] = f
            known.add(f)

# part 2B - execute the stored program
r = [0] * 4
i = max(i for i, x in enumerate(lines) if not x.strip()) + 1
for line in lines[i:]:
    op, a, b, c = parse(line)
    f = opcodes[op]
    f(r, a, b, c)
print(r[0])
